### ISSUE TITLE
Farm implementation?

### ISSUE THREAD

**Issue by RicardoMusch** [cpenv/cpenv#36]
This is more of a question really.

I'm looking at using REZ or CPENV as the current setup of handling packages/plugins and dependencies is getting a bit out of hand. CPENV seems to be easier to use as you've got a ShotGrid integration and thats what we use.

I was wondering, say for example I have CPENV packages for Nuke plugins, once we send a nuke render to the farm, would the NUKE_PATH not now be pointing to localized packages? Normally the nuke plugins live on a shared file server so the NUKE_PATH for the workstation is copied into the Nuke Job and the rendernode looks at the exact same path as Nuke on the workstation does.

It seems I'm a bit confused about how that would influence things?
Any pointers would be appreciated :)

**Comment by danbradham** [comment]
Hey Ricardo,

Sorry for the delayed response.

The general idea is to use cpenv on the farm as well. If you use Deadline, you could use [autocpenv](https://github.com/cpenv/autocpenv). Autocpenv is an event plugin for Deadline that looks for CPENV_ACTIVE_MODULES in the users environment when they submit a job, and passes that along to the job. Then, before a job starts (in globalJobPreload.py) cpenv is called to localize and activate the modules stored in the job on the render node. This way everything runs locally on all workstations and render nodes.

In regards to sharing tools from a network location, I still find it advantageous! I have 1 cpenv module for each application, nuke, maya, etc. These point to the network shares and allow artists to easily share scripts, tools, and gizmos without needing my constant involvement.

For example I've got a module named shared_nuke. The module file looks something like this.
```
name: shared_nuke
description: Points to our shared tools folder on our network for Nuke.
author: Dan
email: dan@email.com
requires: []
environment:
    NUKE_PATH:
        - win: \\network_share\tools\nuke
          mac: /Volumes/network_share/tools/nuke
          linux: /mnt/network_share/tools/nuke
```

In \\network_share\tools\nuke I've got a standard init.py and menu.py file that configure that directory to share gizmos, toolsets, and some plugins. I consider this a kind of best of both worlds approach. The majority of our plugins and tools I deploy as cpenv modules, basically anything that is properly versioned and large enough to warrant it. The shared tools on the network are still there for flexibility, but, they are more manageable because we've got cpenv modules for most of our pivotal pieces.

I should also point out that [rez also has a shotgun integration](https://github.com/nerdvegas/rez-shotgun). The fundamental difference between that integration and cpenv though, is that cpenv stores your Environment configuration on a custom project entity and has a UI to make it easier to work with. While rez is configured in the actual Shotgun config.

tk-cpenv does not have a way of defining default Environments for new jobs, so, for each job you would need to import the Environments from another project via the Set Modules UI. I personally found we have too many jobs being created to keep up with. So I use our Shotgun Event Daemon to copy the Environment entities from the template project when a new project is created.

One last thing, rez and cpenv can be very helpful to maintain software dependencies on a per job basis. But they both require a lot of work. I spend a fair amount of time each month creating cpenv modules when new plugins are released haha. I like it and it's been really cool to see our module list grow as the years have gone on. I've created over 200 at this point, and that number won't stop increasing!

Hopefully I clarified the workflow a bit here, let me know if you have any other questions.

**Comment by RicardoMusch** [comment]
Awesome man! That is a super helpfull and detailed reply.
I've managed to get things working and will soon integrate the farm stuff :)
